### PR TITLE
Member avatars without canvas

### PR DESCRIPTION
**Pull request by clarkf**
Partially addresses https://github.com/vector-im/element-web/issues/23936. Does not address blurhashes/thumbnails or the favicon.

This changes `<BaseAvatar />` to use a span with a `background-color` rather than an `` with a `data:` URL.

This means that user, member and room avatars no longer use canvas-generated colored backgrounds. Firefox users with `resistFingerprinting` enabled will no longer see noise as the background for users and rooms without actual avatars:
image

Note that this does change the actual DOM structure of textual avatars (`img` -> `span`), which means a lot of snapshots changed, bloating this changeset substantially.

## Checklist

-   [x] Tests written for new code (and old code if feasible)
-   [x] Linter and other CI checks pass
-   [x] Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-react-sdk/blob/develop/CONTRIBUTING.md))

---
Here's what your changelog entry will look like:

## 🐛 Bug Fixes
 * Member avatars without canvas ([\#9990](https://github.com/matrix-org/matrix-react-sdk/pull/9990)). Contributed by @clarkf.

### CODE REVIEW

**Comment by andybalaam** [comment]
This is brilliant, thank you. Is the screenshot above the "before" picture? If so, please could you include an "after" picture?

I added the X-Needs-Percy label to run screenshot tests. Before merging I'd like to confirm that we have some screenshot tests that cover this area.

File: src/Avatar.ts [hunk 1]
```diff
@@ -24,16 +24,19 @@ import DMRoomMap from "./utils/DMRoomMap";
 import { mediaFromMxc } from "./customisations/Media";
 import { isLocalRoom } from "./utils/localRoom/isLocalRoom";
 
+const DEFAULT_COLORS: Readonly<string[]> = ["#0DBD8B", "#368bd6", "#ac3ba8"];
```
Comment (andybalaam) [inline src/Avatar.ts:27]: Where did these colours come from?

File: src/components/views/avatars/BaseAvatar.tsx [hunk 2]
```diff
@@ -66,11 +71,26 @@ const calculateUrls = (url: string, urls: string[], lowBandwidth: boolean): stri
...
+    urls,
+}: {
+    url: string | undefined;
+    urls: string[] | undefined;
+}): [string | undefined, () => void] => {
     // Since this is a hot code path and the settings store can be slow, we
     // use the cached lowBandwidth value from the room context if it exists
     const roomContext = useContext(RoomContext);
-    const lowBandwidth = roomContext ? roomContext.lowBandwidth : SettingsStore.getValue("lowBandwidth");
+    const lowBandwidth = roomContext.lowBandwidth;
```
Comment (andybalaam) [inline src/components/views/avatars/BaseAvatar.tsx:93]: Why the change here?

File: src/components/views/avatars/MemberAvatar.tsx [hunk 3]
```diff
@@ -77,8 +76,8 @@ export default function MemberAvatar({
 
         if (!title) {
             title =
-                UserIdentifierCustomisations.getDisplayUserIdentifier(member?.userId ?? "", {
-                    roomId: member?.roomId ?? "",
+                UserIdentifierCustomisations.getDisplayUserIdentifier!(member.userId, {
```
Comment (andybalaam) [inline src/components/views/avatars/MemberAvatar.tsx:79]: This is interesting :-) Why the `!` here?

**Comment by andybalaam** [comment]
My own before and after with Firefox's "resistFingerprinting" config option enabled:

Before:
image

After:
image

**Comment by andybalaam** [comment]
This appears to slightly break the layout of the room header:

After:
image

Before:
image

**Review by andybalaam — CHANGES_REQUESTED** [review]
This change significantly improves the code in several places, thank you!

It seems much more sensible to me to use a span instead of an image here, so all good from my point of view. (Although I guess there may be an historical reason that I am missing?)

A couple of questions from me, but all good.

The remaining thing is to fix the layout of the room header, and check any other places the avatar can appear for similar problems.

We can review the screenshot tests to check they cover things, but I'm not confident they cover everything, so we should check carefully manually.

File: src/Avatar.ts (hunk 1 above)
Comment (clarkf) [reply src/Avatar.ts:27]: https://github.com/matrix-org/matrix-react-sdk/blob/a69b805156126708c78e2a0ed7c5fab3919a7f82/src/Avatar.ts#L91

They were within `defaultAvatarUrlForString`, I just extracted them to a file-level constant.

File: src/components/views/avatars/BaseAvatar.tsx (hunk 2 above)
Comment (clarkf) [reply src/components/views/avatars/BaseAvatar.tsx:93]: `useContext` always returns a value, even if it's just the default. So here `roomContext` is `IRoomState`, not `IRoomState | undefined`.

File: src/components/views/avatars/MemberAvatar.tsx (hunk 3 above)
Comment (clarkf) [reply src/components/views/avatars/MemberAvatar.tsx:79]: This wasn't really central to this change-set, so I just left the behavior what it was. If `getDisplayUserIdentifier` is `undefined`, it would have errored, as it does now.

There's an argument to be made for optional chaining being "more correct". In reality, [`getDisplayUserIdentifier` is *always* defined](https://github.com/matrix-org/matrix-react-sdk/blob/a69b805156126708c78e2a0ed7c5fab3919a7f82/src/customisations/UserIdentifier.ts#L17-L41), but I don't really know enough about the customisation system to be able to say how universal that is.

**Comment by clarkf** [comment]
Good catch on the room header! I'll go through and double check all the instances I can find.

**Comment by clarkf** [comment]
New editor pills:
Before:
image

After:
image

**Comment by clarkf** [comment]
Fixed the instance in `<NewRoomIntro />`:
image

I'm no CSS wiz, so I'm not *entirely* certain why the elements are not growing. Seems like both the outer and the inner `<span>` need to be `display: block` and have an explicit `width`/`height`. ¯\\_(ツ)\_/¯

**Review by andybalaam — APPROVED** [review]
Looks good, thank you!

**Comment by andybalaam** [comment]
@clarkf we hit a whole load of problems caused by this change! Apologies: I should have realised at review time how far-reaching this change is - `BaseAvatar` is mentioned in 70 files within our source repo!

I like the change, but to be able to merge it we would need a more concerted effort:

* You'd need to look at the related bugs and make sure the change does not cause them to happen again: https://github.com/vector-im/element-web/issues/24382 , https://github.com/vector-im/element-web/issues/24370 , https://github.com/vector-im/element-web/issues/24375 , https://github.com/vector-im/element-web/issues/24369 , https://github.com/vector-im/element-web/issues/24374 , https://github.com/vector-im/element-web/issues/24401 , https://github.com/vector-im/element-web/issues/24402 , https://github.com/vector-im/element-web/issues/24409
* We'd need to survey the product and code and get a proper idea of which screens show avatars, and confirm with screenshots that they did not change
* For everywhere it's possible, we'd need to add screenshot tests that prove the change doesn't cause new bugs

This situation was particularly unlucky because at least some of the bugs were caused by a bug in Firefox's rendering, so we'd also need to make sure we tested across the supported browsers.

In summary, this is a much more ambitious change than I realised, and we'd need the requisite level of effort to be able to merge it! Thanks again for your work on this, and if you want to continue do ask for help and we will try to support you.
